src/listpack.c

```diff
@@ -796,6 +822,27 @@ unsigned char *lpAppend(unsigned char *lp, unsigned char *ele, uint32_t size) {
     return lpInsert(lp,ele,size,eofptr,LP_BEFORE,NULL);
 }
 
+/* Append the specified element 's' of length 'slen' at the head of the listpack. */
+unsigned char *lpPushHead(unsigned char *lp, unsigned char *s, uint32_t slen) {
+    unsigned char *p = lpFirst(lp);
+    if (!p) return lpPushTail(lp, s, slen);
+    return lpInsert(lp, s, slen, p, LP_BEFORE, NULL);
+}
+
+/* Append the specified element 's' of length 'slen' at the end of the
+ * listpack. It is implemented in terms of lpInsert(), so the return value is
+ * the same as lpInsert(). */
+unsigned char *lpPushTail(unsigned char *lp, unsigned char *s, uint32_t slen) {
+    uint64_t listpack_bytes = lpGetTotalBytes(lp);
+    unsigned char *eofptr = lp + listpack_bytes - 1;
+    return lpInsert(lp, s, slen, eofptr, LP_BEFORE, NULL);
+}
+
+/* Remove the element pointed by 'p'. */
+unsigned char *lpReplace(unsigned char *lp, unsigned char *p, unsigned char *s, uint32_t slen) {
+    return lpInsert(lp, s, slen, p, LP_REPLACE, NULL);
+}
```

Comment:
i'm trying to understand why you chose to add these adapters here, and the _lpGet in listcontainer.c.
arguably they should have the same fate.
maybe it would be cleaner to promote the declaration of the list container interface to a header of it's own (not in server.h), and then include that header in both ziplist.c and listpack.c and have them create the types (removing the need for listcontainer.c)

Reply:
Great!!! this is exactly the problem I've been struggling with for days (ziplistEntity in ziplist) .